File: RoutinUsage/Providers/NewAPIUsageProvider.swift
import Foundation

struct NewAPIUsageProvider: UsageProvider {
    let descriptor: ProviderDescriptor
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
        self.descriptor = ProviderRegistry.builtInDescriptors.first(where: { $0.id == .newAPI })!
    }

    func validate(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        try await fetchUsage(credential, now: now)
    }

    func fetchUsage(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        guard credential.providerID == .newAPI,
              credential.kind == .bearerAPIKey,
              !credential.secret.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let rawUserID = credential.metadata["userID"],
              let userID = Int(rawUserID.trimmingCharacters(in: .whitespacesAndNewlines)),
              userID > 0,
              let root = Self.baseURL(from: credential.metadata["baseURL"])
        else { throw UsageProviderError.invalidCredential }

        let user: NewAPIUser
        let displayUnit: NewAPIDisplayUnit
        do {
            async let statusResponse = request(
                root.appendingPathComponent("api/status"),
                credential: credential,
                as: NewAPIEnvelope<NewAPIStatus>.self
            )
            async let userResponse = request(
                root.appendingPathComponent("api/user/self"),
                credential: credential,
                as: NewAPIEnvelope<NewAPIUser>.self
            )
            let (statusEnvelope, userEnvelope) = try await (statusResponse, userResponse)
            guard userEnvelope.success != false, let userData = userEnvelope.data else {
                throw UsageProviderError.providerMessage("New API：\(userEnvelope.message ?? "认证失败")")
            }
            user = userData
            displayUnit = NewAPIDisplayUnit(status: statusEnvelope.data)
        } catch let error as UsageProviderError {
            throw error
        } catch {
            throw UsageProviderError.invalidResponse
        }

        let calendar = Calendar.current
        let todayStart = calendar.startOfDay(for: now)
        let oneDayStart = now.addingTimeInterval(-24 * 60 * 60)
        let sevenDayStart = now.addingTimeInterval(-7 * 24 * 60 * 60)
        let thirtyDayStart = now.addingTimeInterval(-30 * 24 * 60 * 60)

        // New API 的个人日志没有时间段 Token 聚合字段；/api/data/self 是日志衍生的
        // 当前用户聚合数据，直接返回 token_used 和 count，可避免分页拉取上万条日志。
        async let todayData = requestTokenSummary(
            root: root,
            start: Self.restrictedStart(todayStart, now: now),
            end: now,
            credential: credential
        )
        async let oneDayData = requestTokenSummary(
            root: root,
            start: Self.restrictedStart(oneDayStart, now: now),
            end: now,
            credential: credential
        )
        async let sevenDayData = requestTokenSummary(
            root: root,
            start: Self.restrictedStart(sevenDayStart, now: now),
            end: now,
            credential: credential
        )
        async let thirtyDayData = requestTokenSummary(
            root: root,
            start: Self.restrictedStart(thirtyDayStart, now: now),
            end: now,
            credential: credential
        )
        async let currentMinuteStat = requestStat(
            root: root,
            start: now.addingTimeInterval(-60),
            end: now,
            credential: credential
        )
        let (
            todaySummary,
            oneDaySummary,
            sevenDaySummary,
            thirtyDaySummary,
            currentMinute
        ) = try await (todayData, oneDayData, sevenDayData, thirtyDayData, currentMinuteStat)

        let used = user.usedQuota
        let remaining = max(0, user.quota)
        let total = used + remaining
        let threshold = credential.metadata["balanceWarningThreshold"]
            .flatMap { Decimal(string: $0) }
            .map { Self.convert($0, using: displayUnit) }
        let convertedRemaining = Self.convert(remaining, using: displayUnit)
        let health = UsageMetricHealthEvaluator.balanceState(
            balance: convertedRemaining,
            warningThreshold: threshold,
            isAvailable: true
        )
        let group = user.group?.trimmingCharacters(in: .whitespacesAndNewlines)
        let planName = group.map { "New API · \($0)" } ?? "New API"

        return UsageSnapshot(
            planName: planName,
            kind: .periodic,
            fiveHour: nil,
            weekly: nil,
            token: nil,
            allowedModels: [],
            fetchedAt: now,
            providerID: .newAPI,
            credentialID: credential.credentialID,
            metrics: [
                NormalizedUsageMetric(
                    id: "quota-progress",
                    label: "账户额度",
                    used: Self.convert(used, using: displayUnit),
                    limit: Self.convert(total, using: displayUnit),
                    remaining: convertedRemaining,
                    unit: .currency,
                    presentation: .progress,
                    currencyCode: displayUnit.symbol,
                    healthState: health
                ),
                NormalizedUsageMetric(
                    id: "today-token",
                    label: "今日 Token",
                    value: Decimal(todaySummary.tokenUsed ?? 0),
                    unit: .token,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "one-day-token",
                    label: "近 24 小时 Token",
                    value: Decimal(oneDaySummary.tokenUsed ?? 0),
                    unit: .token,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "seven-day-token",
                    label: "近 7 天 Token",
                    value: Decimal(sevenDaySummary.tokenUsed ?? 0),
                    unit: .token,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "thirty-day-token",
                    label: "近 30 天 Token",
                    value: Decimal(thirtyDaySummary.tokenUsed ?? 0),
                    unit: .token,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "today-token-cost",
                    label: "今日消费",
                    value: Self.convert(Decimal(todaySummary.quota ?? 0), using: displayUnit),
                    unit: .currency,
                    presentation: .value,
                    currencyCode: displayUnit.symbol,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "one-day-token-cost",
                    label: "近 24 小时消费",
                    value: Self.convert(Decimal(oneDaySummary.quota ?? 0), using: displayUnit),
                    unit: .currency,
                    presentation: .value,
                    currencyCode: displayUnit.symbol,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "seven-day-token-cost",
                    label: "近 7 天消费",
                    value: Self.convert(Decimal(sevenDaySummary.quota ?? 0), using: displayUnit),
                    unit: .currency,
                    presentation: .value,
                    currencyCode: displayUnit.symbol,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "thirty-day-token-cost",
                    label: "近 30 天消费",
                    value: Self.convert(Decimal(thirtyDaySummary.quota ?? 0), using: displayUnit),
                    unit: .currency,
                    presentation: .value,
                    currencyCode: displayUnit.symbol,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "rpm",
                    label: "近 60 秒 RPM",
                    value: Decimal(currentMinute.rpm),
                    unit: .request,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "tpm",
                    label: "近 60 秒 TPM",
                    value: Decimal(currentMinute.tpm),
                    unit: .token,
                    presentation: .value,
                    healthState: .normal
                ),
                NormalizedUsageMetric(
                    id: "request-count",
                    label: "账户累计请求",
                    value: Decimal(user.requestCount),
                    unit: .request,
                    presentation: .value,
                    healthState: .normal
                )
            ]
        )
    }

    private func requestStat(
        root: URL,
        start: Date,
        end: Date,
        credential: ProviderCredential
    ) async throws -> NewAPIStat {
        var components = URLComponents(url: root.appendingPathComponent("api/log/self/stat"), resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "start_timestamp", value: String(Int64(start.timeIntervalSince1970))),
            URLQueryItem(name: "end_timestamp", value: String(Int64(end.timeIntervalSince1970))),
            URLQueryItem(name: "type", value: "2")
        ]
        guard let url = components?.url else { throw UsageProviderError.invalidCredential }
        let envelope = try await request(url, credential: credential, as: NewAPIEnvelope<NewAPIStat>.self)
        guard envelope.success != false, let data = envelope.data else {
            throw UsageProviderError.providerMessage("New API：\(envelope.message ?? "统计接口返回失败")")
        }
        return data
    }

    private func requestTokenSummary(
        root: URL,
        start: Date,
        end: Date,
        credential: ProviderCredential
    ) async throws -> NewAPITokenSummary {
        var components = URLComponents(url: root.appendingPathComponent("api/data/self"), resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "start_timestamp", value: String(Int64(start.timeIntervalSince1970))),
            URLQueryItem(name: "end_timestamp", value: String(Int64(end.timeIntervalSince1970)))
        ]
        guard let url = components?.url else { throw UsageProviderError.invalidCredential }
        let envelope = try await request(url, credential: credential, as: NewAPIEnvelope<[NewAPITokenSummary]>.self)
        guard envelope.success != false else {
            throw UsageProviderError.providerMessage("New API：\(envelope.message ?? "Token 统计接口返回失败")")
        }
        return (envelope.data ?? []).reduce(NewAPITokenSummary()) { partial, item in
            NewAPITokenSummary(
                tokenUsed: (partial.tokenUsed ?? 0) + (item.tokenUsed ?? 0),
                requestCount: (partial.requestCount ?? 0) + (item.requestCount ?? 0),
                quota: (partial.quota ?? 0) + (item.quota ?? 0)
            )
        }
    }

    private static func restrictedStart(_ start: Date, now: Date) -> Date {
        max(start, now.addingTimeInterval(-2592000))
    }

    private func request<T: Decodable>(
        _ url: URL,
        credential: ProviderCredential,
        as type: T.Type
    ) async throws -> T {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        request.setValue("Bearer \(credential.secret)", forHTTPHeaderField: "Authorization")
        request.setValue(credential.metadata["userID"], forHTTPHeaderField: "New-Api-User")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw UsageProviderError.transport
        }
        guard let httpResponse = response as? HTTPURLResponse else {
            throw UsageProviderError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            switch httpResponse.statusCode {
            case 401, 403: throw UsageProviderError.unauthorized
            case 429: throw UsageProviderError.rateLimited
            case 500...599: throw UsageProviderError.providerUnavailable
            default: throw UsageProviderError.invalidResponse
            }
        }
        do {
            return try JSONDecoder().decode(type, from: data)
        } catch {
            throw UsageProviderError.invalidResponse
        }
    }

    private static func baseURL(from rawValue: String?) -> URL? {
        guard let rawValue,
              var components = URLComponents(string: rawValue.trimmingCharacters(in: .whitespacesAndNewlines)),
              components.scheme != nil,
              components.host != nil
        else { return nil }
        while components.path.hasSuffix("/") { components.path.removeLast() }
        if components.path == "/api" { components.path = "" }
        return components.url
    }

    private static func valueMetric(
        id: String,
        label: String,
        value: Decimal,
        displayUnit: NewAPIDisplayUnit,
        health: UsageMetricHealthState
    ) -> NormalizedUsageMetric {
        NormalizedUsageMetric(
            id: id,
            label: label,
            value: value,
            unit: .currency,
            presentation: .value,
            currencyCode: displayUnit.symbol,
            healthState: health
        )
    }

    private static func convert(_ rawQuota: Decimal, using displayUnit: NewAPIDisplayUnit) -> Decimal {
        displayUnit.convert(rawQuota)
    }
}

private struct NewAPIDisplayUnit: Sendable {
    let quotaPerUnit: Decimal
    let exchangeRate: Decimal
    let symbol: String
    let displaysRawQuota: Bool

    init(status: NewAPIStatus?) {
        let quotaPerUnit = status?.quotaPerUnit ?? 500_000
        self.quotaPerUnit = quotaPerUnit > 0 ? quotaPerUnit : 500_000

        let displayType = (status?.quotaDisplayType ?? "USD").uppercased()
        let usdRate = status?.usdExchangeRate ?? 1
        let customRate = status?.customCurrencyExchangeRate ?? 1

        switch displayType {
        case "CNY":
            exchangeRate = usdRate > 0 ? usdRate : 1
            symbol = "¥"
            displaysRawQuota = false
        case "CUSTOM":
            exchangeRate = customRate > 0 ? customRate : 1
            let customSymbol = (status?.customCurrencySymbol ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            symbol = customSymbol.isEmpty ? "单位" : customSymbol
            displaysRawQuota = false
        case "TOKENS":
            exchangeRate = 1
            symbol = "额度"
            displaysRawQuota = true
        default:
            exchangeRate = 1
            symbol = "$"
            displaysRawQuota = false
        }
    }

    func convert(_ rawQuota: Decimal) -> Decimal {
        displaysRawQuota ? rawQuota : rawQuota / quotaPerUnit * exchangeRate
    }
}

private struct NewAPIStatus: Decodable {
    let quotaPerUnit: Decimal?
    let quotaDisplayType: String?
    let usdExchangeRate: Decimal?
    let customCurrencyExchangeRate: Decimal?
    let customCurrencySymbol: String?

    private enum CodingKeys: String, CodingKey {
        case quotaPerUnit = "quota_per_unit"
        case quotaDisplayType = "quota_display_type"
        case usdExchangeRate = "usd_exchange_rate"
        case customCurrencyExchangeRate = "custom_currency_exchange_rate"
        case customCurrencySymbol = "custom_currency_symbol"
    }
}

private struct NewAPITokenSummary: Decodable, Sendable {
    let tokenUsed: Int?
    let requestCount: Int?
    let quota: Int?

    init(tokenUsed: Int? = nil, requestCount: Int? = nil, quota: Int? = nil) {
        self.tokenUsed = tokenUsed
        self.requestCount = requestCount
        self.quota = quota
    }

    private enum CodingKeys: String, CodingKey {
        case tokenUsed = "token_used"
        case requestCount = "count"
        case quota
    }
}

private struct NewAPIEnvelope<Value: Decodable>: Decodable {
    let success: Bool?
    let message: String?
    let data: Value?
}

private struct NewAPIUser: Decodable {
    let quota: Decimal
    let usedQuota: Decimal
    let requestCount: Int
    let group: String?

    private enum CodingKeys: String, CodingKey {
        case quota
        case usedQuota = "used_quota"
        case requestCount = "request_count"
        case group
    }
}

private struct NewAPIStat: Decodable {
    let quota: Decimal
    let rpm: Int
    let tpm: Int
}

File: android/provider/src/main/kotlin/ai/routin/mytoken/provider/newapi/NewAPIUsageProvider.kt
package ai.routin.mytoken.provider.newapi

import ai.routin.mytoken.domain.model.Credential
import ai.routin.mytoken.domain.model.CredentialKind
import ai.routin.mytoken.domain.model.CredentialMetadataKey
import ai.routin.mytoken.domain.model.CredentialSecret
import ai.routin.mytoken.domain.model.ProviderId
import ai.routin.mytoken.domain.model.UsageMetric
import ai.routin.mytoken.domain.model.UsageMetricHealthState
import ai.routin.mytoken.domain.model.UsageMetricPresentation
import ai.routin.mytoken.domain.model.UsageMetricSemantic
import ai.routin.mytoken.domain.model.UsageMetricUnit
import ai.routin.mytoken.domain.model.UsageSnapshot
import ai.routin.mytoken.domain.usage.UsageProvider
import ai.routin.mytoken.domain.usage.UsageProviderException
import ai.routin.mytoken.provider.health.UsageMetricHealthEvaluator
import ai.routin.mytoken.provider.http.HttpTransport
import ai.routin.mytoken.provider.http.ProviderHttpRequest
import ai.routin.mytoken.provider.http.ProviderHttpResponse
import ai.routin.mytoken.provider.json.ProviderJson
import ai.routin.mytoken.provider.json.arrayOrNull
import ai.routin.mytoken.provider.json.asObjectOrNull
import ai.routin.mytoken.provider.json.boolOrNull
import ai.routin.mytoken.provider.json.decimalOrNull
import ai.routin.mytoken.provider.json.intOrNull
import ai.routin.mytoken.provider.json.objOrNull
import ai.routin.mytoken.provider.json.stringOrNull
import kotlinx.serialization.json.JsonObject
import java.math.BigDecimal
import java.math.MathContext
import java.net.URI
import java.net.URLEncoder
import java.time.Clock
import java.time.Instant
import java.time.ZoneId
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * New API adapter. Shares the macOS endpoint contract, including the dashboard
 * envelope (`success/message/data`) and raw quota display conversion.
 */
class NewAPIUsageProvider(
    private val transport: HttpTransport,
    private val clock: Clock = Clock.systemUTC(),
) : UsageProvider {

    override val providerId: ProviderId = ProviderId.NewAPI

    override suspend fun fetchUsage(
        credential: Credential,
        secret: CredentialSecret,
    ): Result<UsageSnapshot> {
        if (credential.providerId != ProviderId.NewAPI || credential.credentialKind != CredentialKind.BearerApiKey) {
            return Result.failure(UsageProviderException.InvalidCredential())
        }
        val token = (secret as? CredentialSecret.BearerToken)?.token.orEmpty()
        val userID = credential.metadata[CredentialMetadataKey.UserID]?.trim().orEmpty()
        val root = parseBaseURL(credential.metadata[CredentialMetadataKey.BaseURL])
        if (token.isBlank() || userID.toLongOrNull() ?: 0L <= 0L || root == null) {
            return Result.failure(UsageProviderException.InvalidCredential())
        }

        return runCatching { fetchSnapshot(credential, token, userID, root) }.recoverCatching { error ->
            if (error is UsageProviderException) throw error
            throw UsageProviderException.InvalidResponse()
        }
    }

    private suspend fun fetchSnapshot(
        credential: Credential,
        token: String,
        userID: String,
        root: URI,
    ): UsageSnapshot {
        val now = clock.instant()
        val todayStart = now.atZone(ZoneId.systemDefault()).toLocalDate()
            .atStartOfDay(ZoneId.systemDefault()).toInstant()
        val statusData = requestObject(credential, token, userID, endpoint(root, "api/status"))
        val userData = requestObject(credential, token, userID, endpoint(root, "api/user/self"))
        requireEnvelopeSuccess(statusData, "状态接口返回失败")
        requireEnvelopeSuccess(userData, "认证失败")
        val user = userData.objOrNull("data") ?: throw UsageProviderException.InvalidResponse()
        val displayUnit = DisplayUnit.fromStatus(statusData.objOrNull("data"))

        return coroutineScope {
            val todayToken = async { tokenSummary(credential, token, userID, root, todayStart, now) }
            val oneDayToken = async { tokenSummary(credential, token, userID, root, now.minusSeconds(86_400), now) }
            val sevenDayToken = async { tokenSummary(credential, token, userID, root, now.minusSeconds(604_800), now) }
            val thirtyDayToken = async { tokenSummary(credential, token, userID, root, now.minusSeconds(2_592_000), now) }
            val minuteStat = async { minuteStat(credential, token, userID, root, now) }
            val today = todayToken.await()
            val oneDay = oneDayToken.await()
            val sevenDay = sevenDayToken.await()
            val thirtyDay = thirtyDayToken.await()
            val stat = minuteStat.await()
            buildSnapshot(credential.id, now, user, stat, today, oneDay, sevenDay, thirtyDay, displayUnit)
        }
    }

    private suspend fun requestObject(
        credential: Credential,
        token: String,
        userID: String,
        url: URI,
    ): JsonObject {
        val response = try {
            transport.execute(
                ProviderHttpRequest(
                    method = "GET",
                    url = url.toString(),
                    headers = mapOf(
                        "Authorization" to "Bearer $token",
                        "New-Api-User" to userID,
                        "Accept" to "application/json",
                    ),
                )
            )
        } catch (error: Exception) {
            if (error is kotlinx.coroutines.CancellationException) throw error
            throw UsageProviderException.Transport()
        }
        val body = response.body.decodeToString()
        if (response.statusCode == 401 || response.statusCode == 403) throw UsageProviderException.Unauthorized()
        if (response.statusCode == 429) throw UsageProviderException.RateLimited()
        if (response.statusCode in 500..599) throw UsageProviderException.ProviderUnavailable()
        if (response.statusCode !in 200..299) throw UsageProviderException.InvalidResponse()
        return ProviderJson.parse(body).asObjectOrNull() ?: throw UsageProviderException.InvalidResponse()
    }

    private suspend fun tokenSummary(
        credential: Credential,
        token: String,
        userID: String,
        root: URI,
        start: Instant,
        end: Instant,
    ): TokenSummary {
        val url = endpoint(
            root,
            "api/data/self",
            "start_timestamp" to start.epochSecond.toString(),
            "end_timestamp" to end.epochSecond.toString(),
        )
        val data = requestObject(credential, token, userID, url)
        requireEnvelopeSuccess(data, "Token 统计接口返回失败")
        return data.arrayOrNull("data").orEmpty().mapNotNull { it.asObjectOrNull() }.fold(TokenSummary()) { total, item ->
            total + TokenSummary(
                tokenUsed = item.intOrNull("token_used") ?: 0,
                requestCount = item.intOrNull("count") ?: 0,
                quota = item.decimalOrNull("quota") ?: BigDecimal.ZERO,
            )
        }
    }

    private suspend fun minuteStat(
        credential: Credential,
        token: String,
        userID: String,
        root: URI,
        now: Instant,
    ): MinuteStat {
        val url = endpoint(
            root,
            "api/log/self/stat",
            "start_timestamp" to now.minusSeconds(60).epochSecond.toString(),
            "end_timestamp" to now.epochSecond.toString(),
            "type" to "2",
        )
        val data = requestObject(credential, token, userID, url)
        requireEnvelopeSuccess(data, "统计接口返回失败")
        val stat = data.objOrNull("data") ?: throw UsageProviderException.InvalidResponse()
        return MinuteStat(
            rpm = stat.intOrNull("rpm") ?: 0,
            tpm = stat.intOrNull("tpm") ?: 0,
        )
    }

    private fun buildSnapshot(
        credentialId: java.util.UUID,
        now: Instant,
        user: JsonObject,
        minute: MinuteStat,
        today: TokenSummary,
        oneDay: TokenSummary,
        sevenDay: TokenSummary,
        thirtyDay: TokenSummary,
        displayUnit: DisplayUnit,
    ): UsageSnapshot {
        val rawUsed = user.decimalOrNull("used_quota") ?: BigDecimal.ZERO
        val rawRemaining = (user.decimalOrNull("quota") ?: BigDecimal.ZERO).max(BigDecimal.ZERO)
        val rawTotal = rawUsed + rawRemaining
        val remaining = displayUnit.convert(rawRemaining)
        val health = UsageMetricHealthEvaluator.balanceState(remaining, null, true)
        val userGroup = user.stringOrNull("group")?.trim().orEmpty()
        val planName = if (userGroup.isEmpty()) "New API" else "New API · $userGroup"

        return UsageSnapshot(
            credentialId = credentialId,
            fetchedAt = now,
            planName = planName,
            metrics = listOf(
                UsageMetric(
                    id = "quota-progress",
                    label = "账户额度",
                    used = displayUnit.convert(rawUsed),
                    limit = displayUnit.convert(rawTotal),
                    remaining = remaining,
                    unit = UsageMetricUnit.Currency,
                    presentation = UsageMetricPresentation.Progress,
                    semantic = UsageMetricSemantic.UsedQuota,
                    currencyCode = displayUnit.symbol,
                    healthState = health,
                ),
                valueMetric("today-token", "今日 Token", today.tokenUsed.toBigDecimal(), UsageMetricUnit.Token),
                valueMetric("one-day-token", "近 24 小时 Token", oneDay.tokenUsed.toBigDecimal(), UsageMetricUnit.Token),
                valueMetric("seven-day-token", "近 7 天 Token", sevenDay.tokenUsed.toBigDecimal(), UsageMetricUnit.Token),
                valueMetric("thirty-day-token", "近 30 天 Token", thirtyDay.tokenUsed.toBigDecimal(), UsageMetricUnit.Token),
                currencyMetric("today-token-cost", "今日消费", today.quota, displayUnit),
                currencyMetric("one-day-token-cost", "近 24 小时消费", oneDay.quota, displayUnit),
                currencyMetric("seven-day-token-cost", "近 7 天消费", sevenDay.quota, displayUnit),
                currencyMetric("thirty-day-token-cost", "近 30 天消费", thirtyDay.quota, displayUnit),
                valueMetric("rpm", "近 60 秒 RPM", minute.rpm.toBigDecimal(), UsageMetricUnit.Request),
                valueMetric("tpm", "近 60 秒 TPM", minute.tpm.toBigDecimal(), UsageMetricUnit.Token),
                valueMetric(
                    "request-count",
                    "账户累计请求",
                    (user.intOrNull("request_count") ?: 0).toBigDecimal(),
                    UsageMetricUnit.Request,
                ),
            ),
        )
    }

    private fun valueMetric(id: String, label: String, value: BigDecimal, unit: UsageMetricUnit) = UsageMetric(
        id = id,
        label = label,
        value = value,
        unit = unit,
        presentation = UsageMetricPresentation.Value,
        semantic = UsageMetricSemantic.Value,
        healthState = UsageMetricHealthState.Normal,
    )

    private fun currencyMetric(id: String, label: String, raw: BigDecimal, unit: DisplayUnit) = UsageMetric(
        id = id,
        label = label,
        value = unit.convert(raw),
        unit = UsageMetricUnit.Currency,
        presentation = UsageMetricPresentation.Value,
        semantic = UsageMetricSemantic.Value,
        currencyCode = unit.symbol,
        healthState = UsageMetricHealthState.Normal,
    )

    private fun requireEnvelopeSuccess(envelope: JsonObject, fallback: String) {
        if (envelope.boolOrNull("success") == false) {
            throw UsageProviderException.ProviderMessage("New API：${envelope.stringOrNull("message") ?: fallback}")
        }
    }

    private data class TokenSummary(
        val tokenUsed: Int = 0,
        val requestCount: Int = 0,
        val quota: BigDecimal = BigDecimal.ZERO,
    ) {
        operator fun plus(other: TokenSummary) = TokenSummary(
            tokenUsed = tokenUsed + other.tokenUsed,
            requestCount = requestCount + other.requestCount,
            quota = quota + other.quota,
        )
    }

    private data class MinuteStat(val rpm: Int, val tpm: Int)

    private data class DisplayUnit(
        val quotaPerUnit: BigDecimal,
        val exchangeRate: BigDecimal,
        val symbol: String,
        val displaysRawQuota: Boolean,
    ) {
        fun convert(rawQuota: BigDecimal): BigDecimal = if (displaysRawQuota) {
            rawQuota
        } else {
            rawQuota.divide(quotaPerUnit, MathContext.DECIMAL128).multiply(exchangeRate)
                .stripTrailingZeros()
        }

        companion object {
            fun fromStatus(status: JsonObject?): DisplayUnit {
                val quotaPerUnit = (status?.decimalOrNull("quota_per_unit") ?: BigDecimal(500_000))
                    .takeIf { it.signum() > 0 } ?: BigDecimal(500_000)
                val displayType = status?.stringOrNull("quota_display_type")?.uppercase().orEmpty()
                return when (displayType) {
                    "CNY" -> DisplayUnit(
                        quotaPerUnit,
                        positiveOr(status?.decimalOrNull("usd_exchange_rate"), BigDecimal.ONE),
                        "¥",
                        false,
                    )
                    "CUSTOM" -> DisplayUnit(
                        quotaPerUnit,
                        positiveOr(status?.decimalOrNull("custom_currency_exchange_rate"), BigDecimal.ONE),
                        status?.stringOrNull("custom_currency_symbol")?.trim().takeUnless { it.isNullOrEmpty() } ?: "单位",
                        false,
                    )
                    "TOKENS" -> DisplayUnit(quotaPerUnit, BigDecimal.ONE, "额度", true)
                    else -> DisplayUnit(quotaPerUnit, BigDecimal.ONE, "$", false)
                }
            }

            private fun positiveOr(value: BigDecimal?, fallback: BigDecimal) =
                value?.takeIf { it.signum() > 0 } ?: fallback
        }
    }

    private companion object {
        fun parseBaseURL(raw: String?): URI? {
            val value = raw?.trim().orEmpty()
            if (value.isEmpty()) return null
            return runCatching { URI(value) }.getOrNull()?.takeIf { it.scheme != null && it.host != null }
        }

        fun endpoint(root: URI, path: String, vararg query: Pair<String, String>): URI {
            val normalizedRoot = root.toString().trimEnd('/')
            val pathSuffix = if (normalizedRoot.endsWith("/api") && path.startsWith("api/")) {
                path.removePrefix("api/")
            } else {
                path
            }
            val encoded = query.joinToString("&") { (key, value) ->
                "${URLEncoder.encode(key, Charsets.UTF_8)}=${URLEncoder.encode(value, Charsets.UTF_8)}"
            }
            return URI("$normalizedRoot/$pathSuffix${if (encoded.isEmpty()) "" else "?$encoded"}")
        }
    }
}
